modules/core/src/parallel.cpp

```diff
@@ -789,12 +826,31 @@ int cv::getNumberOfCPUs(void)
     GetSystemInfo( &sysinfo );
 #endif
 
-    return (int)sysinfo.dwNumberOfProcessors;
+    return concurentThreadsSupported == 0 ? (unsigned)sysinfo.dwNumberOfProcessors : (concurentThreadsSupported < (unsigned)sysinfo.dwNumberOfProcessors ? concurentThreadsSupported : (unsigned)sysinfo.dwNumberOfProcessors);
 #elif defined __ANDROID__
-    static int ncpus = getNumberOfCPUsImpl();
+    static unsigned ncpus_impl = (unsigned)getNumberOfCPUsImpl();
+    static unsigned ncpus = concurentThreadsSupported == 0 ? ncpus_impl : (concurentThreadsSupported < ncpus_impl ? concurentThreadsSupported : ncpus_impl);
+    return ncpus;
+#elif (defined __linux__ || defined __GLIBC__ || defined __HAIKU__ || defined __EMSCRIPTEN__) && !defined __ANDROID__
+    static unsigned ncpus_cgroup = getNumberOfCPUsImplLinux();
+
+    unsigned ncpus_cpu_set = concurentThreadsSupported;
+    if(concurentThreadsSupported == 0) { /* somehow hardware_concurrency failed or c++ version < 11 */
+        cpu_set_t cpu_set;
+        sched_getaffinity(0, sizeof(cpu_set), &cpu_set);
+        ncpus_cpu_set = CPU_COUNT(&cpu_set);
+    }
+
+    /* Cases:
+     * 1. Both are zero return 1
+     * 2. Logical xor of ncpus_cgroup and ncpus_cpu_set is True:
+     *  2.1. ncpus_cgroup is 0, return ncpus_cpu_set
+     *  2.2. ncpus_cpu_set is 0, return ncpus_cgroup
+     * 3. Return min of zoth
+     */
+    static unsigned ncpus = ncpus_cgroup == 0 && ncpus_cpu_set == 0 ? 1 : (!ncpus_cgroup != !ncpus_cpu_set ? ncpus_cgroup | ncpus_cpu_set : (ncpus_cgroup < ncpus_cpu_set ? ncpus_cgroup : ncpus_cpu_set));
```

Comment:
Please refactor very long lines like this one. We have soft limitation of 100 columns.

Reply:
Sure